File: LeetCode/DataStruct/Golang/Leetcode662-maximum-width-of-binary-tree.go
package sbgo

type pair struct {
	node  *TreeNode
	level int
}

func max(i, j int) int {
	if i < j {
		return j
	}
	return i
}
func widthOfBinaryTree(root *TreeNode) int {
	q := []pair{pair{root, 1}}
	mx := 0
	for len(q) != 0 {
		n := len(q)
		l, r := q[0].level, 0
		for i := 0; i < n; i++ {
			t, p := q[0].node, q[0].level-l+1
			r = q[0].level
			q = q[1:]
			if t.Left != nil {
				q = append(q, pair{t.Left, p * 2})
			}
			if t.Right != nil {
				q = append(q, pair{t.Right, p*2 + 1})
			}
		}
		mx = max(mx, r-l+1)
	}
	return mx
}

File: LeetCode/DataStruct/Rust/Leetcode662-maximum-width-of-binary-tree.rs
use std::cell::RefCell;
use std::rc::Rc;
// use std::collections::VecDeque;
impl Solution {
  pub fn width_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
    if root.is_none() {
      return 0;
    }
    let mut ans = 1;
    let mut que = vec![(0, root)];
    while !que.is_empty() {
      ans = ans.max(que.last().unwrap().0 - que[0].0 + 1);
      let mut tmp = vec![];
      for (i, r) in que {
        let r = r.as_ref().unwrap().borrow();
        if r.left.is_some() {
          tmp.push((i * 2, r.left.clone()));
        }
        if r.right.is_some() {
          tmp.push((i * 2 + 1, r.right.clone()));
        }
      }
      que = tmp;
    }
    ans
  }
}
